design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+1. Behave closer to other related Python SDKs (Pinecone, OpenAI, HuggingFace, ...)
+1. Allow infering identity from environment
+1. Provide better abstractions for common, more complex operations (I/O, imports)
+
+## Non-Goals
+
+1. For now, we explicitly leave out environment administration: setting up server, configuring GC rules, all IAM/ACL related things.
+
+## Authentication
+
+Any operation that calls out to lakeFS will try to authenticate using the following chain:
+
+1. All models receive an optional `client` kwarg with explicit credentials
+1. Otherwise, if `lakefs.init(...)` is called with parameters (`access_key_id`, `jwt_token`, ...) - these will be set on a `lakefs.DefaultClient` object
+1. If `lakefs.init()` is called with no parameters, or `init` is not called:
+    1. use `LAKECTL_ACCESS_KEY_ID` and `LAKECTL_ACCESS_SECRET_KEY` if set
+    1. Otherwise, use `~/.lakectl.yaml` if exists
+    1. Otherwise, try and use IAM role from current machine (using AWS IAM role (will only work with enterprise/cloud))
+1. If init is not called, it will be lazily called on the first use of `DefaultClient`, deferring authentication to the first API call.
+
+## API wrapper interface
+
+the higher lavel SDK will be resource-class based. performing API operations will be
+done by calling the methods on their parent object. Examples:
+
+```python
+import lakefs
+
+repo = lakefs.repository('example')
+branch = repo.branch('main')
+for item in branch.list_objects(prefix='foo/'):
+    if item.path.endswith('.parquet'):
+        print(item.path)
+
+data: bytes = branch.download('datasets/foo/1.parquet')
+branch.upload(data, 'foo/bar/baz.parquet')
+
+# this will work:
+data: bytes = lakefs.repository('example').commit('abc123').download('a/b.txt')
+
+# since commits are immutable, this method will not exist:
+lakefs.repository('example').commit('abc123').upload(data, 'a/b.txt')
+```
+
+## Partial interface definition
+
+### Authentication
+
+```python
+class Client:
+    """
+    Wrapper around lakefs_sdk's client object
+    Takes care of instantiating it from the environment
+    """
+    def __init__(self, **kwargs):
+        self._client = _infer_auth_chain(**kwargs)
+
+# global default client
+DefaultClient: Optional[Client] = None
+
+try:
+    DefaultClient = Client()
+except NoAuthenticationFound:
+    # must call init() explicitly
+    DefaultClient = None
+
+
+def init(**kwargs):
+    global DefaultClient
+    DefaultClient = Client(**kwargs)
+
+```
+
+### model driven interface
+
+
+
+```python
+class Repository:
+    def __init__(self, repository_id: str, client: Client = DefaultClient): ...
+    def get_or_create(self, storage_namespace: str, default_branch_id: str = 'main', include_samples: bool = False) -> Repository: ...
+    def branch(self, branch_id: str) -> Branch: ...
+    def list_branches(self, max_amount: Optional[int], after: str ='', prefix: str ='') -> Generator[Branch]: ...
+    def commit(self, commit_id: str) -> Reference: ...
```

Comment:
Confusing nomenclature: "commit" makes me expect a verb.  We could `get_commit` and `commit` rather than `commit` and `commit_changes`, but that just means someone else will be confused.

I suggest separating the API into read / write / list operators:
```py
ref = repo.get.commit(branch_or_tag_or_ref_expr)
with r = ref.get.download(...):
  pass

for tag in repo.list.tags(prefix=...):
  pass

branch = repo.get.branch(source_branch).create.branch(branch_name)
branch.create.upload(...)
with r = branch.get.download(...):
  pass
branch.create.commit(message, user_metadata)
```
I prefer to have actual `SUBJECT.list.OBJECT` than `SUBJECT.list_OBJECT` because then I get a nice class diagra and more re-use: the types of `branch.list` and of `ref.list` are the same, the types of `branch.get` and `ref.get` are the same, and `branch.create` has a subtype of `ref.create`.

Reply:
That's clever! Not sure how idiomatic this is in Python though. Let's try and find examples!

Another suggestion: differentiate entities from actions using case: Entities, like classes are [CapWords](https://peps.python.org/pep-0008/#class-names), while actions are methods and are [lowercase_with_underscode](https://peps.python.org/pep-0008/#function-and-variable-names) (a-la [PEP8](https://peps.python.org/pep-0008)). 

### Examples:

We can take a more classic object-oriented approach, returning an object of said entity. This is similar to what [Pinecone is doing](https://docs.pinecone.io/docs/python-client#index) in their API:


```python
repo = lakefs.Repository('example-repo')
ref = repo.Commit('branch_or_tag_or_ref_expr')

with ref.download(...) as f:
  pass

for tag in repo.list_tags(prefix='...'):
  pass

branch = repo.Branch(source_branch).create(branch_name)
branch.upload(...)

with branch.download(...) as r:
  pass

branch.commit(message, user_metadata)
```

Alternatively we can use resources as more of a namespace. This is what [OpenAI typically does](https://platform.openai.com/docs/api-reference/files/create?lang=python) with its Python SDK:

```python
c = lakefs.Commit.get(repository='', id='')
b = lakefs.Branch.get(repository='', id='')

for b in lakefs.Branch.list(repository=''):
    print(b)

for c in lakefs.Commit.log(repository='', starting_at=''):
    print(c)


with lakefs.Object.open(repository='', reference='', path='', mode='r') as f:
    f.read()


for p in lakefs.Object.list(repository='', reference='', prefix=''):
    print(p)
```

Which eliminates all ambiguity but means the caller has to pass all context (repository, ref, etc) on every call, which is verbose and possibly tedious. 

That being said, I don't oppose the suggestions you gave at all! 
I do want to make sure we're doing something that Python users will find familiar - let's find examples!

Reply:
So I searched for some CRDL examples (I switched the "U" for an "L").  There seems to be more than one way to do it.

* SQAlchemy [does something similar](https://docs.sqlalchemy.org/en/14/orm/tutorial.html) to both of our suggestions I think.  Probably a bit closer to yours.
* K8s [mostly copies the Go syntax](https://github.com/kubernetes-client/python/blob/master/examples/ingress_create.py).  It seems to give the complete REST experience in Python.  I don't think that we want to go there, people who want that can already use `lakefs_sdk`.
* Boto is most like your Pinecone example, here's [an arbitrary sample](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/collections.html).  But you get an actual constructor(-ish), I think.

The advantage of Boto is that it is hands-down the largest API.  So we know that something like this can grow, and that AWS may have _already_ seen many possible issues.  OTOH I am not sure of its Pythonicity.

Boto would be more like this, I think:
```py
c = lakefs.Repository('example').Commit(Id='')
b = lakefs.Repository('example').Branch(Id='')

for b in lakefs.Repository('example').branches.all()
    print(b)

for c in lakefs.Repository(...).Commit(...).log()   # log gives a generator
    print(c)

# open cannot take a mode, we should split into open and create.  Otherwise it
# is a really late dynamic error when you try to Repository.Ref.Object.create
# instead of Repository.Branch.Object.create.
with lakefs.Repository('example').Ref('branch').Object('key').open() as f:
    f.read()
# I could easily be convinced to allow shortcuts to skip over some or all of
# the qualifiers above, to allow these:
with lakefs.Repository(repository='example', branch='branch', key='key').create(...):
    pass
with lakefs.Object(repository='example', branch='branch', key='key').create(...):
    pass

for p in lakefs.Repository(...).Ref(...).list(prefix=..., ...):
    print(p)
```

Reply:
@arielshaqed PTAL - I tried to adapt it to the Boto style - I believe it's clearer but would appreciate your review.